src/inc/til/u8u16convert.h

```diff
@@ -306,25 +89,117 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
     // Arguments:
     // - in - UTF-8 string to be converted
     // - out - reference to the resulting UTF-16 string
-    // - state - reference to a til::u8state class holding the status of the current partials handling
+    // - state - reference to a til::u8state holding the status of the current partials handling
     // Return Value:
     // - S_OK          - the conversion succeeded
     // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
     // - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
     // - E_UNEXPECTED  - an unexpected error occurred
-    template<class inT, class outT>
-    [[nodiscard]] typename std::enable_if<std::is_same<typename inT::value_type, char>::value && std::is_same<typename outT::value_type, wchar_t>::value, HRESULT>::type
-    u8u16(const inT in, outT& out, u8state& state) noexcept
+    template<class outT>
+    [[nodiscard]] HRESULT u8u16(const std::string_view& in, outT& out, u8state& state) noexcept
     {
-        std::string_view sv{};
-        RETURN_IF_FAILED(state(std::string_view{ in }, sv));
-        return til::u8u16(sv, out);
+        try
+        {
+            out.clear();
+            RETURN_HR_IF(S_OK, in.empty());
+
+            int capa16{};
+            // The worst ratio of UTF-8 code units to UTF-16 code units is 1 to 1 if UTF-8 consists of ASCII only.
+            RETURN_HR_IF(E_ABORT, !base::CheckAdd(in.length(), state.have).AssignIfValid(&capa16));
+
+            out.resize(gsl::narrow_cast<size_t>(capa16));
+            auto len8{ gsl::narrow_cast<int>(in.length()) };
+            int len16{};
+            auto cursor8{ in.data() };
+            if (state.have)
+            {
+                if (state.want > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials + state.have);
+                    state.have += gsl::narrow_cast<uint8_t>(len8);
+                    state.want -= gsl::narrow_cast<uint8_t>(len8);
+                    out.clear();
+                    return S_OK;
+                }
+
+                std::move(cursor8, cursor8 + state.want, state.partials + state.have);
+                len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials, gsl::narrow_cast<int>(state.have) + state.want, out.data(), capa16);
+                RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                capa16 -= len16;
+                len8 -= state.want;
+                cursor8 += state.want;
+                state.reset();
+            }
+
+            if (len8)
+            {
+                auto backIter{ cursor8 + len8 };
+                // If the last byte in the string was a byte belonging to a UTF-8 multi-byte character
+                if (backIter[-1] & 0b1'0000000)

```

Comment:
BTW Tangential to the other comment-thread we have: Would it make sense to remove this if condition and only handle case where the trailing sequence is a valid partial sequence?
That way we could remove some of the if conditions all together... Something roughly like:
```cpp
auto backIter = cursor8 + len8 - 1;
int sequenceLen = 1;

// Skip UTF8 continuation bytes
while (backIter != cursor8 && (*backIter & 0b11'000000) == 0b10'000000)
{
    ++sequenceLen;
}

if ((*backIter & 0b11'000000) == 0b11'000000)
{
    static constexpr uint8_t lengths[] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 2, 3, 3, 4, 0};
    const auto codePointLen = lengths[ch >> 3];

    if (codePointLen > sequenceLen)
    {
        std::move(backIter, backIter + sequenceLen, state.partials);
        len8 -= sequenceLen;
        state.have = gsl::narrow_cast<uint8_t>(sequenceLen);
        state.want = gsl::narrow_cast<uint8_t>(codePointLen - sequenceLen);
    }
}
```

Unless I did a off-by-one error or something similar this should properly work for all given inputs but is a bit less code and more linear.
It doesn't have to stop early after 3 bytes IMO as performance doesn't necessarily need to be guaranteed for vastly invalid inputs. This code should still be sufficiently fast (~3GB/s for backwards byte-wise iterations like that).

Reply:
len8 still has to be >0 to avoid an off-by-one. So, we can't skip this check. It should actually work as expected as soon as we include the decrementation of backiter in the while loop.

Reply:
Ah sorry, the code I posted was meant to be inside the `if (len8)` condition.
Thanks for noticing the missing decrement!

Reply:
I have to thank you for taking the time for all of your kind reviews!